File: Wire-iOS/Sources/Authentication/RegistrationPersonal/Steps/NoHistoryViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "NoHistoryViewController.h"
@import WireSyncEngine;
#import "RegistrationFormController.h"
#import "Button.h"
#import "Wire-Swift.h"

@interface NoHistoryViewController ()
@property (nonatomic) UILabel *heroLabel;
@property (nonatomic) UILabel *subtitleLabel;
@property (nonatomic) UIView *contentView;
@property (nonatomic) UIStackView *stackView;
@property (nonatomic) UIImageView *backgroundImageView;
@end

@implementation NoHistoryViewController

@synthesize authenticationCoordinator;

- (instancetype)initWithContextType:(ContextType)contextType
{
    self = [super initWithNibName:nil bundle:nil];
    if (nil != self) {
        _contextType = contextType;
    }
    return self;
}
    
- (void)viewDidLoad
{
    [super viewDidLoad];
    [self createBackgroundImageView];
    [self createContentView];
    [self createHeroLabel];
    [self createSubtitleLabel];
    
    [self createContentViewConstraints];
    
    [self createButtons];
    
    // Layout first to avoid the initial layout animation during the presentation. 
    [self.view layoutIfNeeded];
}

- (void)createBackgroundImageView
{
    UIImage *backgroundImage = [UIImage imageNamed:@"LaunchImage"];
    self.backgroundImageView = [[UIImageView alloc] initWithImage:backgroundImage];
    self.backgroundImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.backgroundImageView];

    [NSLayoutConstraint activateConstraints:
    @[
          [self.backgroundImageView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
          [self.backgroundImageView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
          [self.backgroundImageView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
          [self.backgroundImageView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
    ]];
}

- (void)createContentView
{
    self.contentView = [[UIView alloc] init];
    self.contentView.translatesAutoresizingMaskIntoConstraints = NO;
    self.contentView.backgroundColor = [UIColor clearColor];
    
    self.view.backgroundColor = [UIColor clearColor];
    
    [self.view addSubview:self.contentView];
 
    self.stackView = [[UIStackView alloc] init];
    self.stackView.translatesAutoresizingMaskIntoConstraints = NO;
    self.stackView.spacing = 12;
    self.stackView.distribution = UIStackViewDistributionFill;
    self.stackView.axis = UILayoutConstraintAxisVertical;
    self.stackView.alignment = UIStackViewAlignmentFill;
    self.stackView.backgroundColor = [UIColor clearColor];
    [self.contentView addSubview:self.stackView];
}

- (void)createHeroLabel
{
    self.heroLabel = [[UILabel alloc] init];
    self.heroLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.heroLabel.numberOfLines = 0;
    self.heroLabel.textColor = [[ColorScheme defaultColorScheme] colorWithName:ColorSchemeColorTextForeground
                                                                       variant:ColorSchemeVariantDark];
    
    NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.paragraphSpacing = 10;
    NSString *heroText = [self localizableStringForPart:@"hero"];
    NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:NSLocalizedString(heroText, nil)
                                                                                       attributes:@{ NSParagraphStyleAttributeName : paragraphStyle,
                                                                                                     NSFontAttributeName: UIFont.largeSemiboldFont }];
    self.heroLabel.attributedText = attributedText;
    [self.stackView addArrangedSubview:self.heroLabel];
}

- (void)createSubtitleLabel
{
    self.subtitleLabel = [[UILabel alloc] init];
    self.subtitleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.subtitleLabel.numberOfLines = 0;
    self.subtitleLabel.textColor = [[ColorScheme defaultColorScheme] colorWithName:ColorSchemeColorTextForeground
                                                                           variant:ColorSchemeVariantDark];
    NSMutableParagraphStyle *paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
    paragraphStyle.paragraphSpacing = 10;
    NSString *subtitleText = [self localizableStringForPart:@"subtitle"];
    NSMutableAttributedString *attributedText = [[NSMutableAttributedString alloc] initWithString:NSLocalizedString(subtitleText, nil)
                                                                                       attributes:@{ NSParagraphStyleAttributeName : paragraphStyle,
                                                                                                     NSFontAttributeName: UIFont.largeThinFont}];
    self.subtitleLabel.attributedText = attributedText;
    [self.stackView addArrangedSubview:self.subtitleLabel];
    
    UIView *placeholder = [[UIView alloc] init];
    placeholder.backgroundColor = [UIColor clearColor];
    placeholder.translatesAutoresizingMaskIntoConstraints = NO;
    [placeholder.heightAnchor constraintEqualToConstant:24];
    [self.stackView addArrangedSubview:placeholder];
}

- (NSString *)localizableStringForPart:(NSString *)part
{
    switch (self.contextType) {
        case ContextTypeNewDevice:
            return [@"registration.no_history" stringByAppendingPathExtension:part];
            break;
        case ContextTypeLoggedOut:
            return [@"registration.no_history.logged_out" stringByAppendingPathExtension:part];
            break;
    }
}

@end

File: Wire-iOS/Sources/Authentication/RegistrationPersonal/Steps/NoHistoryViewController.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireDataModel

extension NoHistoryViewController {
    static let WireBackupUTI = "com.wire.backup-ios"
    
    @objc public func createButtons() {
        let restoreBackupButton = Button(style: .emptyMonochrome)
        restoreBackupButton.translatesAutoresizingMaskIntoConstraints = false
        restoreBackupButton.setTitle("registration.no_history.restore_backup".localized.uppercased(), for: .normal)
        
        restoreBackupButton.addCallback(for: .touchUpInside) { [showWarningMessage] _ in
            showWarningMessage()
        }
        
        restoreBackupButton.heightAnchor.constraint(equalToConstant: 40).isActive = true
        stackView.addArrangedSubview(restoreBackupButton)
    
        let okButton = Button(style: .fullMonochrome)
        okButton.translatesAutoresizingMaskIntoConstraints = false
        let gotItText = self.localizableString(forPart: "got_it")!
        okButton.setTitle(gotItText.localized, for: .normal)
        okButton.addCallback(for: .touchUpInside) { [unowned self] _ in
            self.authenticationCoordinator?.completeBackupStep()
        }
        
        okButton.heightAnchor.constraint(equalToConstant: 40).isActive = true
        stackView.addArrangedSubview(okButton)
    }
    
    @objc public func createContentViewConstraints() {
        if self.traitCollection.horizontalSizeClass == .regular {
            NSLayoutConstraint.activate([
                contentView.widthAnchor.constraint(equalToConstant: parent!.maximumFormSize.width),
                contentView.heightAnchor.constraint(equalToConstant: parent!.maximumFormSize.height),
                contentView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
                contentView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
            ])
        }
        else {
            contentView.fitInSuperview()
        }
        
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 28),
            stackView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -28 - UIScreen.safeArea.bottom),
            stackView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -28)
        ])
    }
    
    fileprivate func showFilePicker() {
        // Test code to verify restore
        #if arch(i386) || arch(x86_64)
            let testFilePath = "/var/tmp/backup.ios_wbu"
            if FileManager.default.fileExists(atPath: testFilePath) {
                self.restore(with: URL(fileURLWithPath: testFilePath))
                return
            }
        #endif
        
        let picker = UIDocumentPickerViewController(documentTypes: [NoHistoryViewController.WireBackupUTI], in: .`import`)
        picker.delegate = self
        self.present(picker, animated: true)
    }
    
    fileprivate func restore(with url: URL) {
        requestPassword { [performRestore] password in
            performRestore(password, url)
        }
    }
    
    fileprivate func performRestore(using password: String, from url: URL) {
        guard let sessionManager = SessionManager.shared else { return }
        spinnerView.subtitle = "registration.no_history.restore_backup.restoring".localized.uppercased()
        showLoadingView = true
        
        sessionManager.restoreFromBackup(at: url, password: password) { [weak self] result in
            guard let `self` = self else { return }
            switch result {
            case .failure(SessionManager.BackupError.decryptionError):
                self.showLoadingView = false
                self.showWrongPasswordAlert {
                    self.restore(with: url)
                }
            case .failure(let error):
                BackupEvent.importFailed.track()
                self.showRestoreError(error)
                self.showLoadingView = false
            case .success:
                BackupEvent.importSucceeded.track()
                self.spinnerView.subtitle = "registration.no_history.restore_backup.completed".localized.uppercased()
                self.indicateLoadingSuccessRemovingCheckmark(false) {
                    self.authenticationCoordinator?.completeBackupStep()
                }
            }
        }
    }
    
    // MARK: - Alerts
    
    fileprivate func showWarningMessage() {
        let controller = UIAlertController.historyImportWarning { [showFilePicker] in
            showFilePicker()
        }
        present(controller, animated: true)
    }
    
    fileprivate func requestPassword(completion: @escaping (String) -> Void) {
        let controller = UIAlertController.requestRestorePassword { password in
            password.apply(completion)
        }
        present(controller, animated: true, completion: nil)
    }
    
    fileprivate func showWrongPasswordAlert(completion: @escaping () -> Void) {
        let controller = UIAlertController.importWrongPasswordError(completion: completion)
        present(controller, animated: true, completion: nil)
    }
    
    fileprivate func showRestoreError(_ error: Error) {
        let controller = UIAlertController.restoreBackupFailed(with: error) { [unowned self] action in
            switch action {
            case .tryAgain: self.showFilePicker()
            case .cancel: self.authenticationCoordinator?.completeBackupStep()
            }
        }
        present(controller, animated: true)
    }
}

extension NoHistoryViewController: UIDocumentPickerDelegate {
    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentAt url: URL) {
        self.restore(with: url)
    }
}
